### PR TITLE
pass http requests arguments to app documents

### LINKED ISSUE #4828: Feature: Make GET arguments available for bokeh server apps
GET arguments are an important feature of the web, so I think they should be available for server apps. 
They are complementary to session variables (#3349) and also mentioned [here](https://github.com/bokeh/bokeh/wiki/Use-of-the-query-string-in-url-to-control-plot-properties), but for the purpose of plot configuration.

I already have [working code](https://github.com/bkueng/bokeh/commit/3dfe97f6ad64e79f7658c9393a1afd290a4361f4) for it, but as per guidelines I open an issue first.

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #4828
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

@havocp I would appreciate your feedback on this.  If there are request arguments like the session id, those are included in the results passed to doc. Does that seem OK? Is `curdoc` the best place to put these? Is there a reason to have a mode that does not pass request args?

### CODE REVIEW

File: bokeh/server/application_context.py [hunk 1]
```diff
@@ -137,7 +137,7 @@ def run_unload_hook(self):
         self.server_context._remove_all_callbacks()
 
     @gen.coroutine
-    def create_session_if_needed(self, session_id):
+    def create_session_if_needed(self, session_id, request_arguments=None):
```
Comment (bryevdv) [inline bokeh/server/application_context.py]: This same function is called by the WS view handler, which has no request arguments to pass, hence the default `None` parameter. Not sure if there's a cleaner way to go here.

**Comment by havocp** [comment]
can't look in depth until Monday but a quick question did you consider putting these args on the SessionContext? If using the lifecycle hooks to get these is inconvenient, maybe for convenience the document could have doc.session_context which would be None for docs that aren't on the server side. So doc.session_context.request_args. That would also provide access to any future session attributes (and the session id or whatever attributes it already has).

A potential point of confusion here may be that the args are always for the session-creating request, not any subsequent request using the same session.  But most of the time that shouldn't matter (?).

**Comment by SiggyF** [comment]
I implemented the suggested changes (`request_args` in `session_context` and `session_context` in `document`). I have implemented them in [a branch](https://github.com/SiggyF/bokeh/tree/fedor/4828_get_args) in my fork. I'm not sure how to merge it into this pull request.

**Comment by SiggyF** [comment]
@bryevdv, I added a pull request that extends on this one #4912. If you accept that pull request, the changes should be merged in this one. It contains:
- adding request_args as part of the session_context
- adding the session_context to the document.

**Comment by bryevdv** [comment]
@SiggyF That's great, I just got back from a short holiday and am a bit buried but I will take a close look at this in the next few days.

@havocp can you think of any reasons no to just expose the entire tornado request?

**Comment by havocp** [comment]
I thought about that a little. The one worry I have is that we would hold a reference to the request for the lifetime of the app. I don't know how large it is or whether this would matter. 

Otherwise it's appealing to go ahead and give people all the info we have.

**Comment by bryevdv** [comment]
@SiggyF do you have any insight on that question? I don't know offhand. My offhand expectation is that it is not a large object but I don't know for sure.

**Comment by SiggyF** [comment]
The size of the `request`, according to `pympler.asizeof`, is 25kB. I have updated my PR #4912 , which can be merged with this PR. It now exposes the full request (as a read-only property).

**Comment by bryevdv** [comment]
@havocp any final comments? Will merge soon if not

**Comment by bryevdv** [comment]
OK I am going to add a short docs section then merge this today.

**Comment by bryevdv** [comment]
Will merge when Travis is green

**Pull request merged**